pkg/trace/obfuscate/sql.go

```diff
@@ -195,20 +200,41 @@ type tableFinderFilter struct {
 // Filter implements tokenFilter.
 func (f *tableFinderFilter) Filter(token, lastToken TokenKind, buffer []byte) (TokenKind, []byte, error) {
 	switch lastToken {
-	case From:
+	case From, Join:
 		// SELECT ... FROM [tableName]
 		// DELETE FROM [tableName]
+		// ... JOIN [tableName]
 		if r, _ := utf8.DecodeRune(buffer); !unicode.IsLetter(r) {
 			// first character in buffer is not a letter; we might have a nested
 			// query like SELECT * FROM (SELECT ...)
 			break
 		}
 		fallthrough
-	case Update, Into, Join:
+	case Update, Into:
 		// UPDATE [tableName]
 		// INSERT INTO [tableName]
-		// ... JOIN [tableName]
-		f.storeName(string(buffer))
+		if f.storeNames {
+			f.storeName(string(buffer))
+		}
+
+		if f.normalizeTables && token == ID {
+			// Replace all digits
+			buf := make([]byte, 0, len(buffer))
+			scanningDigit := false
+			for _, c := range string(buffer) {
```

Comment:
If this is going to run on _every_ table name, we should aim to improve it. Are we not able to achieve the same outcome without creating a new byte slice and using the same one instead?

Reply:
The problem is we have to iterate over characters, not bytes. Which means we have to decode the utf8 chars. I tried to reuse the same byte buffer but it ended up being a lot worse:

```
BenchmarkTableNormalizer/alloc-8         	20158231	        65.2 ns/op	      32 B/op	       1 allocs/op
BenchmarkTableNormalizer/inplace-8       	 3155025	       382 ns/op	     152 B/op	       4 allocs/op
```

Here is the code:

```go
func (f *replaceFilter) replaceTableDigitsInPlace(buffer []byte) []byte {
	i := 0 // buffer read position
	j := 0 // buffer write position
	scanningDigit := false

	for i < len(buffer) {
		r, size := utf8.DecodeRune(buffer[i : len(buffer)-1])
		i += size

		if r == utf8.RuneError {
			log.Debugf("UTF-8 decoding error when normalizing table digits for table: %s", string(buffer))
			break
		}

		if unicode.IsDigit(r) {
			if scanningDigit {
				continue
			}
			scanningDigit = true
			buffer[j] = byte('?')
			j++
			continue
		}
		scanningDigit = false
		j += utf8.EncodeRune(buffer[j:len(buffer)-1], r)
	}
	return buffer[0:j]
}
```